### PR TITLE
Quorum support for ISemaphore

### PR DESCRIPTION
**Pull request by tombujok**
Operations not marked as mutating:

- SemaphoreReplicationOperation
- SemaphoreDetachMemberOperation
- SemaphoreDetachMemberBackupOperation

EE side: https://github.com/hazelcast/hazelcast-enterprise/pull/1805

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/concurrent/semaphore/SemaphoreQuorumWriteTest.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    public static void setUp() {
+        initTestEnvironment(new Config(), new TestHazelcastInstanceFactory());
+    }
+
+    @AfterClass
+    public static void tearDown() {
+        shutdownTestEnvironment();
+    }
+
+    // init(permits)
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/concurrent/semaphore/SemaphoreQuorumWriteTest.java]: leftover notes?

File: hazelcast/src/test/java/com/hazelcast/concurrent/semaphore/AbstractSemaphoreQuorumTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            QuorumConfig quorumConfig = newQuorumConfig(quorumType, quorumName);
+            SemaphoreConfig semaphoreConfig = newSemaphoreConfig(quorumType, quorumName);
+            config.addQuorumConfig(quorumConfig);
+            config.addSemaphoreConfig(semaphoreConfig);
+            quorumNames[i++] = quorumName;
+        }
+
+        CLUSTER.createFiveMemberCluster(config);
+        for (QuorumType quorumType : types) {
+            ISemaphore semaphore = CLUSTER.instance[0].getSemaphore(SEMAPHORE + quorumType.name());
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/concurrent/semaphore/AbstractSemaphoreQuorumTest.java]: maybe it's possible to abstract away the data-structure-specific parts to reduce code duplication in other datastructures' quorum tests?

Comment (tombujok) [reply hazelcast/src/test/java/com/hazelcast/concurrent/semaphore/AbstractSemaphoreQuorumTest.java]: Will try at the end, but the lack of closures makes it a bit pointless.

File: hazelcast/src/test/java/com/hazelcast/concurrent/semaphore/SemaphoreQuorumWriteTest.java (hunk 1 above)
Comment (tombujok) [reply hazelcast/src/test/java/com/hazelcast/concurrent/semaphore/SemaphoreQuorumWriteTest.java]: Yep, will delete. Thx

**Pull request merged**